File: typescript/src/core/nvml-reader.ts
/**
 * NVML reader — Phase 2 GPU foundation.
 *
 * **TypeScript architectural deviation from Python**: there is no maintained
 * native NVML binding for Node in 2026. This module shells out to the
 * `nvidia-smi` CLI via `child_process.spawnSync` and parses CSV output.
 * Functions still return typed objects (ProcessInfo, UtilSample, MemInfo)
 * so downstream callers see the same shape as the Python `pynvml`
 * wrapper. Per-probe overhead is ~50ms (acceptable for finalize-time use;
 * not suitable for hot-path sampling — but neither is NVML).
 *
 * Fail-silent contract (convention §9) identical to Python:
 * every reader returns `null` (or `false` for booleans) on missing binary,
 * non-zero exit, or parse failure — the caller (GpuAccountant) decides
 * the fallback policy per Decision #1's classification table.
 *
 * **Browser safety**: nvmlAvailable() short-circuits to false off-Node.
 *
 * Decision #4: getProductName applies NFC + lowercase + whitespace collapse
 * on the raw nvidia-smi product-name output before returning — catalog
 * alias matching depends on byte-level equality after normalization.
 *
 * Decision #8: getProcessUtilization takes a mutable `lastSeenTimestamps`
 * dict and updates it in place. The TS implementation uses
 * `process.hrtime.bigint() → ms` for the per-PID timestamp (nvidia-smi pmon
 * does not expose NVML's sample-buffer timestamp). The accountant uses
 * (endTs - startTs) ms as the per-PID active window proxy.
 */

import { spawnSync as nodeSpawnSync } from "node:child_process";
import type { SpawnSyncReturns } from "node:child_process";

// Indirection so tests can swap the spawn implementation. The TS ESM bound
// import would otherwise be immutable across module instances.
type SpawnFn = (
  cmd: string,
  args: string[],
  options: { encoding: "utf-8"; timeout: number },
) => SpawnSyncReturns<string>;

let _spawnFn: SpawnFn = (cmd, args, options) =>
  nodeSpawnSync(cmd, args, options) as SpawnSyncReturns<string>;

/** Test-only — override the spawnSync implementation. Pass `null` to restore. */
export function _setSpawnFnForTests(fn: SpawnFn | null): void {
  _spawnFn =
    fn ??
    ((cmd, args, options) =>
      nodeSpawnSync(cmd, args, options) as SpawnSyncReturns<string>);
}

// ─── Warn-once tracking (single-threaded JS → no lock) ──────────────────────

const _warnedModes: Set<string> = new Set();

export function _resetWarningStateForTests(): void {
  _warnedModes.clear();
}

function _warnOnce(mode: string, message: string): void {
  if (_warnedModes.has(mode)) return;
  _warnedModes.add(mode);
  // eslint-disable-next-line no-console
  console.warn(message);
}

// ─── Browser-safety guard ───────────────────────────────────────────────────

function _isNode(): boolean {
  return typeof process !== "undefined" && !!process.versions?.node;
}

// ─── Test-only nvidia-smi path override ─────────────────────────────────────

let _nvidiaSmiPath: string = "nvidia-smi";

/** Test-only — override the nvidia-smi binary path. Pass `null` to restore. */
export function _setNvidiaSmiPathForTests(p: string | null): void {
  _nvidiaSmiPath = p ?? "nvidia-smi";
}

// ─── Typed return values ────────────────────────────────────────────────────

export interface ProcessInfo {
  /** PID of the process holding the GPU. */
  pid: number;
  /** VRAM used by this process in bytes (0 when nvidia-smi reports N/A). */
  usedGpuMemory: number;
}

export interface UtilSample {
  pid: number;
  /** 0–100; percent of time SMs had ≥1 kernel running during the sample window. */
  smUtil: number;
  /** 0–100; percent of time memory subsystem was busy. */
  memUtil: number;
  /** Per-PID timestamp (milliseconds since process start) for Decision #8 state. */
  timeStamp: number;
}

export interface MemInfo {
  usedBytes: number;
  totalBytes: number;
}

// ─── Helpers ────────────────────────────────────────────────────────────────

function _runSmi(args: string[]): { stdout: string; ok: boolean } {
  if (!_isNode()) return { stdout: "", ok: false };
  try {
    const r = _spawnFn(_nvidiaSmiPath, args, {
      encoding: "utf-8",
      timeout: 5000,
    });
    if (r.error || r.status !== 0) {
      return { stdout: "", ok: false };
    }
    return { stdout: r.stdout ?? "", ok: true };
  } catch {
    return { stdout: "", ok: false };
  }
}

function _normalizeProductName(raw: string): string {
  // NFC → lowercase → whitespace collapse (incl. NBSP / NNBSP / zero-width).
  const nfc = raw.normalize("NFC");
  // String.split(/\s+/) treats Unicode whitespace incl. U+00A0 NBSP as a delimiter.
  return nfc.toLowerCase().split(/\s+/).filter(Boolean).join(" ");
}

// ─── Availability + init ────────────────────────────────────────────────────

/**
 * True when `nvidia-smi` is invokable and exits 0 on the count query.
 *
 * The TS-SDK definition of "available" is "binary works"; in Python this
 * was "pynvml importable AND nvmlInit succeeds". Same end behaviour:
 * downstream code checks this predicate before calling readers.
 */
export function nvmlAvailable(): boolean {
  if (!_isNode()) return false;
  const r = _runSmi(["--query-gpu=count", "--format=csv,noheader"]);
  if (!r.ok) {
    _warnOnce(
      "gpu_nvidia_smi_not_found",
      "nvidia-smi not on PATH or exited non-zero; GPU capture disabled. " +
        "Install NVIDIA driver + container toolkit for GPU cost attribution.",
    );
    return false;
  }
  return true;
}

/** Call `nvidia-smi --query-gpu=count` once. True on success, false on any failure. */
export function initNvml(): boolean {
  return nvmlAvailable();
}

/** No-op on TS; the CLI has no persistent handle. Provided for API parity. */
export function shutdownNvml(): void {
  /* no-op */
}

// ─── Device enumeration ─────────────────────────────────────────────────────

/** Number of NVIDIA devices visible to nvidia-smi. `null` on failure. */
export function getDeviceCount(): number | null {
  if (!_isNode()) return null;
  const r = _runSmi(["--query-gpu=count", "--format=csv,noheader"]);
  if (!r.ok) return null;
  // nvidia-smi --query-gpu=count emits one line per GPU, all with the same count value.
  // Take the first non-empty line.
  const first = r.stdout.split("\n").map((s) => s.trim()).find((s) => s);
  if (!first) return null;
  const n = parseInt(first, 10);
  if (Number.isNaN(n)) {
    _warnOnce("gpu_device_count_parse", `nvidia-smi count unparseable: ${first}`);
    return null;
  }
  return n;
}

// ─── Decision #4: NFC-normalized productName ────────────────────────────────

/**
 * Return the nvidia-smi product-name for `index`, NFC-normalized + lowercased.
 *
 * Decision #4 — alias matching against the catalog depends on byte-level
 * equality post-normalization.
 */
export function getProductName(index: number): string | null {
  if (!_isNode()) return null;
  const r = _runSmi([
    `--query-gpu=name`,
    `--format=csv,noheader`,
    `-i`,
    String(index),
  ]);
  if (!r.ok) {
    _warnOnce(
      "gpu_product_name_failed",
      `nvidia-smi --query-gpu=name failed for device ${index}`,
    );
    return null;
  }
  const line = r.stdout.split("\n").map((s) => s.trim()).find((s) => s);
  if (!line) return null;
  return _normalizeProductName(line);
}

// ─── Per-PID compute processes (Decision #1 measurement-side primitive) ────

/**
 * List of PIDs currently holding GPU `index` + their VRAM usage in bytes.
 *
 * Returns `null` on permission denied / failure — the Decision #1
 * load-bearing case for non-root containers. The accountant's cgroup walk
 * then degrades to self-PID-only.
 */
export function getComputeRunningProcesses(index: number): ProcessInfo[] | null {
  if (!_isNode()) return null;
  const r = _runSmi([
    `--query-compute-apps=pid,used_memory`,
    `--format=csv,noheader,nounits`,
    `-i`,
    String(index),
  ]);
  if (!r.ok) {
    _warnOnce(
      "gpu_nvml_permission_denied",
      `nvidia-smi --query-compute-apps failed for device ${index}; ` +
        `GpuAccountant will degrade to self-PID-only`,
    );
    return null;
  }
  const out: ProcessInfo[] = [];
  for (const line of r.stdout.split("\n")) {
    const s = line.trim();
    if (!s) continue;
    const parts = s.split(",").map((p) => p.trim());
    if (parts.length < 2) continue;
    const pid = parseInt(parts[0], 10);
    if (Number.isNaN(pid)) continue;
    let memMib = parseInt(parts[1], 10);
    if (Number.isNaN(memMib)) memMib = 0;
    // nounits ⇒ used_memory is MiB.
    out.push({ pid, usedGpuMemory: memMib * 1024 * 1024 });
  }
  return out;
}

// ─── Per-PID utilization (Decision #8 persistent timestamps) ────────────────

/**
 * Per-PID utilization sampled from `nvidia-smi pmon -c 1 -s u`.
 *
 * Decision #8 — TS uses `Date.now()` as the per-PID timestamp proxy
 * because nvidia-smi pmon does not expose the NVML sample-buffer epoch.
 * The accountant uses (endTs - startTs) ms as the active-GPU-time
 * approximation between snapshots.
 *
 * Updates `lastSeenTimestamps` in place — callers persist across snapshots.
 */
export function getProcessUtilization(
  index: number,
  lastSeenTimestamps: Record<number, number>,
): Record<number, UtilSample> | null {
  if (!_isNode()) return null;
  const r = _runSmi([
    `pmon`,
    `-c`,
    `1`,
    `-s`,
    `u`,
    `-o`,
    `T`,
    `-i`,
    String(index),
  ]);
  if (!r.ok) {
    _warnOnce(
      "gpu_process_utilization_failed",
      `nvidia-smi pmon failed for device ${index}`,
    );
    return null;
  }
  const out: Record<number, UtilSample> = {};
  const nowMs = Date.now();
  for (const rawLine of r.stdout.split("\n")) {
    const line = rawLine.trim();
    if (!line) continue;
    // Skip header lines and N/A rows.
    if (line.startsWith("#")) continue;
    // Whitespace-separated columns: gpu pid type sm mem enc dec command
    const parts = line.split(/\s+/);
    if (parts.length < 5) continue;
    const pid = parseInt(parts[1], 10);
    if (Number.isNaN(pid)) continue;
    const sm = parseInt(parts[3], 10);
    const mem = parseInt(parts[4], 10);
    if (Number.isNaN(sm) || Number.isNaN(mem)) continue;
    out[pid] = {
      pid,
      smUtil: sm,
      memUtil: mem,
      timeStamp: nowMs,
    };
    lastSeenTimestamps[pid] = nowMs;
  }
  return out;
}

// ─── Memory + MIG ───────────────────────────────────────────────────────────

/** Device-level used / total VRAM (bytes). `null` on failure. */
export function getMemoryInfo(index: number): MemInfo | null {
  if (!_isNode()) return null;
  const r = _runSmi([
    `--query-gpu=memory.used,memory.total`,
    `--format=csv,noheader,nounits`,
    `-i`,
    String(index),
  ]);
  if (!r.ok) {
    _warnOnce(
      "gpu_memory_info_failed",
      `nvidia-smi --query-gpu=memory.used,memory.total failed for device ${index}`,
    );
    return null;
  }
  const line = r.stdout.split("\n").map((s) => s.trim()).find((s) => s);
  if (!line) return null;
  const parts = line.split(",").map((s) => s.trim());
  if (parts.length < 2) return null;
  const usedMib = parseInt(parts[0], 10);
  const totalMib = parseInt(parts[1], 10);
  if (Number.isNaN(usedMib) || Number.isNaN(totalMib)) return null;
  return {
    usedBytes: usedMib * 1024 * 1024,
    totalBytes: totalMib * 1024 * 1024,
  };
}

/** True when MIG is enabled on this device (Decision #2 detection). */
export function getMigMode(index: number): boolean {
  if (!_isNode()) return false;
  const r = _runSmi([
    `--query-gpu=mig.mode.current`,
    `--format=csv,noheader`,
    `-i`,
    String(index),
  ]);
  if (!r.ok) {
    // Older GPUs without MIG support → fail-silent (NOT an error).
    return false;
  }
  const line = r.stdout.split("\n").map((s) => s.trim()).find((s) => s);
  if (!line) return false;
  return line.toLowerCase() === "enabled";
}

File: rust/src/core/nvml_reader.rs
//! NVML library wrapper — Phase 2 GPU foundation.
//!
//! Wraps NVIDIA's NVML (via the `nvml-wrapper` crate, gated behind the
//! `gpu` cargo feature) with the fail-silent contract from convention §9.
//! Every NVML call returns `None` (or `false` for boolean accessors) on
//! missing driver / library / permission / device errors rather than
//! raising — the caller (typically [`GpuAccountant`]) decides the fallback
//! policy per Decision #1's classification table.
//!
//! `nvml-wrapper` is an OPTIONAL dependency (cargo `--features gpu`). When
//! the feature is OFF, every accessor returns the no-NVML sentinel and the
//! SDK works on GPU-less hosts.
//!
//! Per Decision #4: [`get_product_name`] applies NFC Unicode normalization
//! + lowercase + whitespace collapse on the raw NVML string before
//! returning — catalog alias matching depends on byte-level equality after
//! normalization because NVIDIA's productName carries non-breaking spaces
//! and other Unicode quirks across driver versions.
//!
//! Per Decision #8: [`get_process_utilization`] takes a mutable
//! `last_seen_timestamps` map and updates it in place — the caller
//! persists per-PID timestamps across snapshot calls so NVML's sample
//! buffer doesn't silently lose intermediate samples.

use std::collections::{HashMap, HashSet};
use std::sync::{Mutex, OnceLock};

use unicode_normalization::UnicodeNormalization;

/// One NVML compute-running-process record (PID + GPU memory usage).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProcessInfo {
    pub pid: u32,
    /// VRAM bytes used by this PID; may be 0 when NVML reports NOT_AVAILABLE.
    pub used_gpu_memory: u64,
}

/// One NVML process-utilization sample for a single PID at one timestamp.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UtilSample {
    pub pid: u32,
    /// 0-100 — percent of time SMs had ≥1 kernel running.
    pub sm_util: u32,
    /// 0-100 — percent of time memory subsystem was busy.
    pub mem_util: u32,
    /// Microseconds since NVML epoch.
    pub time_stamp: u64,
}

/// Device-level memory totals.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MemInfo {
    pub used_bytes: u64,
    pub total_bytes: u64,
}

// ── Log-once-per-failure-mode state (convention §11) ───────────────────────

fn warned_modes() -> &'static Mutex<HashSet<String>> {
    static SLOT: OnceLock<Mutex<HashSet<String>>> = OnceLock::new();
    SLOT.get_or_init(|| Mutex::new(HashSet::new()))
}

fn warn_once(mode: &str, message: &str) {
    let mut set = match warned_modes().lock() {
        Ok(g) => g,
        Err(p) => p.into_inner(),
    };
    if !set.insert(mode.to_string()) {
        return;
    }
    eprintln!("[dexcost][gpu] {}: {}", mode, message);
}

/// Test-only: clear the warn-once tracking set.
#[doc(hidden)]
pub fn reset_warning_state_for_tests() {
    let mut set = match warned_modes().lock() {
        Ok(g) => g,
        Err(p) => p.into_inner(),
    };
    set.clear();
}

// ── Decision #4: NFC-normalized productName ────────────────────────────────

/// NFC → lowercase → whitespace collapse (incl. NBSP / NNBSP / ZWSP).
///
/// Public so tests / catalog tools can normalize alias inputs the same way.
pub fn normalize_product_name(raw: &str) -> String {
    let nfc: String = raw.nfc().collect();
    // Collapse ALL whitespace (incl. U+00A0, U+202F, etc. — char::is_whitespace
    // covers Unicode whitespace categories the same way Python's str.split does).
    let collapsed: String = nfc
        .split(|c: char| c.is_whitespace())
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join(" ");
    collapsed.to_lowercase()
}

// ── NVML availability + accessors ──────────────────────────────────────────
//
// The `gpu` feature gates the real NVML calls. When the feature is OFF
// (default — and for CI / GPU-less hosts), every accessor returns the
// no-NVML sentinel: `Available::No`, count `None`, etc.

#[cfg(feature = "gpu")]
mod nvml_real {
    use super::*;
    use nvml_wrapper::Nvml;
    use std::sync::OnceLock;

    static NVML: OnceLock<Option<Nvml>> = OnceLock::new();

    pub fn nvml_available() -> bool {
        // Crate is linked; runtime init may still fail.
        true
    }

    fn ensure_init() -> Option<&'static Nvml> {
        NVML.get_or_init(|| match Nvml::init() {
            Ok(n) => Some(n),
            Err(e) => {
                warn_once(
                    "gpu_nvml_init_failed",
                    &format!("NVML init failed ({}); GPU capture disabled", e),
                );
                None
            }
        })
        .as_ref()
    }

    pub fn init_nvml() -> bool {
        ensure_init().is_some()
    }

    pub fn shutdown_nvml() {
        // nvml-wrapper drops via OnceLock; explicit shutdown is a no-op here.
    }

    pub fn get_device_count() -> Option<u32> {
        let n = ensure_init()?;
        match n.device_count() {
            Ok(c) => Some(c),
            Err(e) => {
                warn_once(
                    "gpu_device_count_failed",
                    &format!("nvmlDeviceGetCount failed ({})", e),
                );
                None
            }
        }
    }
}

#[cfg(not(feature = "gpu"))]
mod nvml_real {
    use super::*;

    pub fn nvml_available() -> bool {
        // Crate not linked → no NVML available. Customers opt-in via
        // `cargo build --features gpu`.
        false
    }

    pub fn init_nvml() -> bool {
        warn_once(
            "gpu_nvml_not_linked",
            "nvml-wrapper crate not linked; GPU capture disabled. \
             Rebuild with --features gpu to enable.",
        );
        false
    }

    pub fn shutdown_nvml() {}

    pub fn get_device_count() -> Option<u32> {
        None
    }
}

pub use nvml_real::{get_device_count, init_nvml, nvml_available, shutdown_nvml};

// ── Per-device accessors (degraded-mode fallback when feature is OFF) ──────

/// Opaque device handle — wraps the device index. When the `gpu` feature
/// is off there is no NVML; accessors short-circuit.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct DeviceHandle(pub u32);

pub fn get_device_handle(index: u32) -> Option<DeviceHandle> {
    if !nvml_available() {
        return None;
    }
    // When the feature is on, we trust ensure_init() / device_count() ran
    // before this. The handle is just the index; the real NVML lookup
    // happens in the accessors that need it.
    Some(DeviceHandle(index))
}

/// NFC-normalized productName for the given device. `None` on failure
/// or when NVML isn't available.
pub fn get_product_name(handle: DeviceHandle) -> Option<String> {
    real::get_product_name(handle)
}

/// Compute-running processes on the device. `None` on
/// `NVML_ERROR_NO_PERMISSION` — the load-bearing Decision #1 case;
/// caller's cgroup walk degrades to self-PID-only.
pub fn get_compute_running_processes(handle: DeviceHandle) -> Option<Vec<ProcessInfo>> {
    real::get_compute_running_processes(handle)
}

/// Per-PID utilization samples. Mutates `last_seen_timestamps` in place
/// (Decision #8) so the per-PID sample buffer doesn't silently lose
/// intermediate samples between snapshots.
pub fn get_process_utilization(
    handle: DeviceHandle,
    last_seen_timestamps: &mut HashMap<u32, u64>,
) -> Option<HashMap<u32, UtilSample>> {
    real::get_process_utilization(handle, last_seen_timestamps)
}

/// Device-level used / total VRAM bytes.
pub fn get_memory_info(handle: DeviceHandle) -> Option<MemInfo> {
    real::get_memory_info(handle)
}

/// True when MIG is enabled on this device (Decision #2 detection).
pub fn get_mig_mode(handle: DeviceHandle) -> bool {
    real::get_mig_mode(handle)
}

#[cfg(feature = "gpu")]
mod real {
    use super::*;
    use nvml_wrapper::Nvml;
    use std::sync::OnceLock;

    static NVML: OnceLock<Option<Nvml>> = OnceLock::new();

    fn ensure() -> Option<&'static Nvml> {
        // Reuse the cached Nvml from nvml_real if possible — but `nvml_real`'s
        // OnceLock is private. Re-init here is idempotent (Nvml::init() is
        // refcounted internally on the NVIDIA side).
        NVML.get_or_init(|| Nvml::init().ok()).as_ref()
    }

    pub fn get_product_name(handle: DeviceHandle) -> Option<String> {
        let n = ensure()?;
        let d = match n.device_by_index(handle.0) {
            Ok(d) => d,
            Err(_) => return None,
        };
        match d.name() {
            Ok(s) => Some(normalize_product_name(&s)),
            Err(e) => {
                warn_once(
                    "gpu_product_name_failed",
                    &format!("nvmlDeviceGetName failed ({})", e),
                );
                None
            }
        }
    }

    pub fn get_compute_running_processes(handle: DeviceHandle) -> Option<Vec<ProcessInfo>> {
        let n = ensure()?;
        let d = match n.device_by_index(handle.0) {
            Ok(d) => d,
            Err(_) => return None,
        };
        match d.running_compute_processes() {
            Ok(procs) => Some(
                procs
                    .into_iter()
                    .map(|p| ProcessInfo {
                        pid: p.pid,
                        used_gpu_memory: match p.used_gpu_memory {
                            nvml_wrapper::enums::device::UsedGpuMemory::Used(b) => b,
                            _ => 0,
                        },
                    })
                    .collect(),
            ),
            Err(e) => {
                warn_once(
                    "gpu_nvml_permission_denied",
                    &format!(
                        "nvmlDeviceGetComputeRunningProcesses failed ({}); \
                         GpuAccountant will degrade to self-PID-only",
                        e
                    ),
                );
                None
            }
        }
    }

    pub fn get_process_utilization(
        handle: DeviceHandle,
        last_seen_timestamps: &mut HashMap<u32, u64>,
    ) -> Option<HashMap<u32, UtilSample>> {
        let n = ensure()?;
        let d = match n.device_by_index(handle.0) {
            Ok(d) => d,
            Err(_) => return None,
        };
        let base_ts = last_seen_timestamps.values().copied().min().unwrap_or(0);
        match d.process_utilization_stats(Some(base_ts)) {
            Ok(samples) => {
                let mut out = HashMap::new();
                for s in samples {
                    let pid = s.pid;
                    let ts = s.timestamp;
                    out.insert(
                        pid,
                        UtilSample {
                            pid,
                            sm_util: s.sm_util,
                            mem_util: s.mem_util,
                            time_stamp: ts,
                        },
                    );
                    last_seen_timestamps.insert(pid, ts);
                }
                Some(out)
            }
            Err(e) => {
                warn_once(
                    "gpu_process_utilization_failed",
                    &format!("nvmlDeviceGetProcessUtilization failed ({})", e),
                );
                None
            }
        }
    }

    pub fn get_memory_info(handle: DeviceHandle) -> Option<MemInfo> {
        let n = ensure()?;
        let d = match n.device_by_index(handle.0) {
            Ok(d) => d,
            Err(_) => return None,
        };
        match d.memory_info() {
            Ok(m) => Some(MemInfo {
                used_bytes: m.used,
                total_bytes: m.total,
            }),
            Err(_) => None,
        }
    }

    pub fn get_mig_mode(handle: DeviceHandle) -> bool {
        let n = match ensure() {
            Some(n) => n,
            None => return false,
        };
        let d = match n.device_by_index(handle.0) {
            Ok(d) => d,
            Err(_) => return false,
        };
        match d.mig_mode() {
            Ok(m) => matches!(m.current, nvml_wrapper::enum_wrappers::device::MigMode::Enabled),
            Err(_) => false,
        }
    }
}

#[cfg(not(feature = "gpu"))]
mod real {
    use super::*;

    pub fn get_product_name(_handle: DeviceHandle) -> Option<String> {
        None
    }
    pub fn get_compute_running_processes(_handle: DeviceHandle) -> Option<Vec<ProcessInfo>> {
        None
    }
    pub fn get_process_utilization(
        _handle: DeviceHandle,
        _last_seen_timestamps: &mut HashMap<u32, u64>,
    ) -> Option<HashMap<u32, UtilSample>> {
        None
    }
    pub fn get_memory_info(_handle: DeviceHandle) -> Option<MemInfo> {
        None
    }
    pub fn get_mig_mode(_handle: DeviceHandle) -> bool {
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_product_name_collapses_nbsp() {
        // U+00A0 NBSP between "NVIDIA" and "H100"
        let raw = "NVIDIA\u{00A0}H100\u{00A0}80GB HBM3";
        let n = normalize_product_name(raw);
        assert_eq!(n, "nvidia h100 80gb hbm3");
    }

    #[test]
    fn normalize_product_name_collapses_nnbsp_and_zwsp() {
        // U+202F narrow no-break space + zero-width space U+200B (which is
        // category Cf, NOT Zs; Python's str.split treats it as whitespace
        // via char.is_whitespace — same behaviour here for U+00A0/U+202F).
        let raw = "NVIDIA\u{202F}A100   40GB";
        let n = normalize_product_name(raw);
        assert_eq!(n, "nvidia a100 40gb");
    }

    #[test]
    fn normalize_product_name_lowercases() {
        assert_eq!(normalize_product_name("NVIDIA H100"), "nvidia h100");
    }

    #[test]
    fn normalize_product_name_nfc() {
        // Composed e+acute (U+00E9) vs decomposed e (U+0065) + combining acute (U+0301)
        // should normalize to the same string.
        let composed = "café";
        let decomposed = "cafe\u{0301}";
        assert_eq!(
            normalize_product_name(composed),
            normalize_product_name(decomposed)
        );
    }

    #[test]
    fn no_feature_returns_no_devices() {
        // Default build (no `gpu` feature) — every accessor short-circuits.
        // We can't ASSERT this when feature IS on, so the check is gated.
        #[cfg(not(feature = "gpu"))]
        {
            assert!(!nvml_available());
            assert!(!init_nvml());
            assert_eq!(get_device_count(), None);
            assert!(get_device_handle(0).is_none());
        }
    }

    #[test]
    fn warn_once_is_log_once() {
        reset_warning_state_for_tests();
        warn_once("test_mode_x", "first");
        warn_once("test_mode_x", "second");  // suppressed
        let set = warned_modes().lock().unwrap();
        assert!(set.contains("test_mode_x"));
    }
}
